auth/postgres/repo.go

```diff
@@ -0,0 +1,558 @@
+// Copyright (c) Abstract Machines
+// SPDX-License-Identifier: Apache-2.0
+
+package postgres
+
+import (
+	"context"
+	"time"
+
+	"github.com/absmach/supermq/auth"
+	"github.com/absmach/supermq/pkg/errors"
+	repoerr "github.com/absmach/supermq/pkg/errors/repository"
+	"github.com/absmach/supermq/pkg/postgres"
+	"github.com/lib/pq"
+)
+
+var _ auth.PATSRepository = (*patRepo)(nil)
+
+const (
+	saveQuery = `
+	INSERT INTO pats (
+		id, user_id, name, description, secret, issued_at, expires_at, 
+		updated_at, last_used_at, revoked, revoked_at
+	) VALUES (
+		:id, :user_id, :name, :description, :secret, :issued_at, :expires_at,
+		:updated_at, :last_used_at, :revoked, :revoked_at
+	)`
+
+	retrieveQuery = `
+		SELECT 
+		id, user_id, name, description, secret, issued_at, expires_at,
+		updated_at, last_used_at, revoked, revoked_at
+		FROM pats WHERE user_id = $1 AND id = $2`
+
+	saveScopeQuery = `
+		INSERT INTO pat_scopes (pat_id, platform_type, domain_id, domain_type, operation_type, entity_ids)
+		VALUES (:pat_id, :platform_type, :domain_id, :domain_type, :operation_type, :entity_ids)`
+
+	updateScopeQuery = `
+		UPDATE pat_scopes SET
+			platform_type = :platform_type, 
+			domain_id = :domain_id, 
+			domain_type = :domain_type, 
+			operation_type = :operation_type, 
+			entity_ids = :entity_ids
+		WHERE pat_id = :pat_id`
+
+	retrieveScopesQuery = `
+		SELECT platform_type, domain_id, domain_type, operation_type, entity_ids
+		FROM pat_scopes WHERE pat_id = $1`
+
+	deleteScopesQuery = `DELETE FROM pat_scopes WHERE pat_id = $1`
+)
+
+type patRepo struct {
+	db    postgres.Database
+	cache auth.Cache
+}
+
+func NewPatRepo(db postgres.Database, cache auth.Cache) auth.PATSRepository {
+	return &patRepo{
+		db:    db,
+		cache: cache,
+	}
+}
+
+func (pr *patRepo) Save(ctx context.Context, pat auth.PAT) error {
+	record, scope, err := patToDBRecords(pat)
+	if err != nil {
+		return errors.Wrap(repoerr.ErrCreateEntity, err)
+	}
+
+	row, err := pr.db.NamedQueryContext(ctx, saveQuery, record)
+	if err != nil {
+		return postgres.HandleError(repoerr.ErrCreateEntity, err)
+	}
+	defer row.Close()
+
+	scopeRow, err := pr.db.NamedQueryContext(ctx, saveScopeQuery, scope)
+	if err != nil {
+		return postgres.HandleError(repoerr.ErrCreateEntity, err)
+	}
+	defer scopeRow.Close()
+
+	if err := pr.cache.Save(ctx, pat.ID, pat); err != nil {
+		return errors.Wrap(repoerr.ErrCreateEntity, err)
+	}
```

Comment:
Please User transaction(tx), 
If anyone fails then all others should be reverted. With Tx you can achieve it easily, For refer implmenetation, please check roles, clients, channels repo.